gms/gossiper.cc

```diff
@@ -1614,12 +1611,14 @@ future<> gossiper::real_mark_alive(inet_address addr) {
 
     // Make a copy for endpoint_state because the code below can yield
     endpoint_state state = local_state;
-    co_await update_live_endpoints_version();
+    auto lock = co_await lock_endpoint_update_semaphore();
+    co_await replicate_live_endpoints_on_change();
     if (_endpoints_to_talk_with.empty()) {
         _endpoints_to_talk_with.push_back({addr});
     } else {
         _endpoints_to_talk_with.front().push_back(addr);
     }
+    lock.return_all();
```

Comment:
How about using a scope instead?
```cpp
{
    auto lock = co_await lock_endpoint_update_semaphore();
    co_await replicate_live_endpoints_on_change();
    if (_endpoints_to_talk_with.empty()) {
        _endpoints_to_talk_with.push_back({addr});
    } else {
        _endpoints_to_talk_with.front().push_back(addr);
    }
}
```

(I see you use a scope in the other place)

Reply:
Next version introduces a mutate function that acquires the lock, bumps the version if needed and calls replicate...on_change